packages/react-client/src/hooks/useScreenShare.ts

```diff
@@ -49,17 +59,29 @@ export const useScreenShare = (): ScreenshareApi => {
     await Promise.all(addTrackPromises);
   };
 
+  const cleanMiddleware = useCallback(() => {
+    cleanMiddlewareFnRef.current?.();
+    cleanMiddlewareFnRef.current = null;
+  }, []);
+
   const setTracksMiddleware = async (middleware: TracksMiddleware | null): Promise<void> => {
     if (!state?.stream) return;
 
-    const [videoTrack, audioTrack] = getTracks(state.stream);
-    const [newVideoTrack, newAudioTrack] = middleware?.(videoTrack, audioTrack) ?? [videoTrack, audioTrack];
+    const [currentVideo, currentAudio] = getTracks(state.stream);
+
+    cleanMiddleware();
+
+    if (middleware) {
+      const { videoTrack, audioTrack, onClear } = middleware(currentVideo, currentAudio);
+      cleanMiddlewareFnRef.current = onClear;
+      await replaceTracks(videoTrack, audioTrack);
+    }
 
-    await replaceTracks(newVideoTrack, newAudioTrack);
+    await replaceTracks(currentVideo, currentAudio);
```

Comment:
nit: this implementation could look like one in `startStreaming`. 
```ts
let [video, audio] = getTracks(stream);

if (middleware) {
  const { videoTrack, audioTrack, onClear } = middleware(video, audio);
  video = videoTrack;
  audio = audioTrack;
  cleanMiddlewareFnRef.current = onClear;
}

await replaceTracks(video, audio);
```

Or simplify it even more:

```
const [currentVideo, currentAudio] = getTracks(state.stream);

cleanMiddleware();

const { videoTrack, audioTrack, onClear } = middleware 
  ? middleware(currentVideo, currentAudio)
  : { videoTrack: currentVideo, audioTrack: currentAudio, onClear: null }
 
cleanMiddlewareFnRef.current = onClear;
await replaceTracks(videoTrack, audioTrack);
```